### ISSUE TITLE
[RFC] Proposal to use yaml for scraping definitions

### ISSUE THREAD

**Issue by mrwoofa** [stashapp/stash#244]
As some of you are aware, I'm running my own API for metadata (https://metadataapi.net/) - this scrapes a bunch of sites and imports the found metadata for usage.

All my scrapers are written in PHP currently, but i was thinking that if we have them as Yaml files, then anyone can write one and they are portable. This would allow a language-agnostic approach to writing scrapers.

Some sites require some advanced scripting to gather information, so we may not be able to scrape every single site, but we could do most.

I've written a sample Definition for the Porn Pros network. It consists of 3 sections - list, scenes and performers.

List is the configuration for gathering a list of scenes to scrape, usually from the search page.

Scene is to scrape a single scene page

Performer is to scrape a single performer page

I've added some ideas like having a list of selectors so you could try multiple, along with the attribute to grab. Also add Regex filtering, to remove unwanted information.

I currently have 52 Scrapers in my source code, but it's not very portable due to being a bunch of PHP Classes.

Here is a sample definition for Porn Pros https://gist.github.com/mrwoofa/40a596a9e2f0513246bc4a52b889056f

Let me know what you think

**Comment by WithoutPants** [comment]
I'm probably going to need a more thorough schema explanation to work this out.

Also suggest taking a look at the [scraping configuration wiki page](https://github.com/stashapp/stash/wiki/Scraping-configuration) (which is kind of outdated with #236 ) and apply some stuff from there into the schema. Specifically:
- URLs are specified so that the UI can match an entered URL to a configured scraper and perform a scrape. Need a way to do this in your schema for performers and scenes.
- We provide a way to query performers using a provided name. Would be nice to add the required fields to the schema to do this.
- Need a way to run scripts instead of doing the scraping directly; basically backwards compatibility with the existing behaviour.

Alternatively, the YAML scraping configuration could live separately to the JSON scraping configuration, and just have the JSON reference the YAML file, but that doubles the number of files we'd need, and it'd be nice to have it in one file.

**Comment by mrwoofa** [comment]
The schema provided is just a sample, it will change and be improved as we add more sites, as some sites require more complex filtering and scraping. 

But the basis is each datapoint (title, description, date, image) have their own scraping definition, and in that you define a Xpath selector as a string for defaults, or you can define the selector, attribute (href,src) and filters like regex search/replace.

My idea for this definition was for scraping entire sites into a database, i didn't really think about it for one of pages. While those one of pages do work, there are some sites where the data required is only available on the search page, so would be easier adding in bulk.

We could add a section defining what url points to what, so you can scrape a single page. This is obviously up for discussion.

Using the search page to find performers/scenes was out of scope for my idea, as my idea was to scrape entire sites, but can be easily added to work for both..

I chose Yaml because it's the simplest config file format for doing what we need, that most languages can read natively. JSON is hard good for programming, but hard to read for Humans. The idea was to make it easy for even novices to modify. It's quite simple to convert between JSON and Yaml, so we could provide both.

**Comment by WithoutPants** [comment]
The following is the first iteration for the format, which is just converting the existing JSON into YAML and adapting it towards where we need it to go:

```
name: IAFD
performerByName:
  - action: script
    script: 
      - python
        iafdScrape.py
        query
performerByFragment:
  - action: script
    script:
      - python
        iafdScrape.py
        scrape
performerByURL:
  - url: iafd.com
    action: script
    script:
      - python
        iafdScrape.py
        scrapeURL
```

It defines scraping rules for three different scrape types:
- `performerByName` is for searching by name in the performer edit page, returning performer fragments
- `performerByFragment` is for searching for a complete performer record passing in a performer fragment. Intended as a following to the `performerByName` scrape, after the user has selected a performer from the result
- `performerByURL` is for scraping from a URL, entered in the performer edit page. 
All of these are mentioned in a little more detail in the wiki page.

This is a port of the IAFD example config I came up with in the original performer scraper PR.

The following is a more advanced scraper configuration for PornHub. It is intended to perform the scrape within the stash code, instead of via a script. It contains everything I anticipate needing after porting from the python script, but it is untested, so may need further iterations. I've annotated this one to hopefully make it easier to understand:

```
# name of the scraper
name: PornHub
# scraping rules for scraping performer by URL
performerByURL:
  # the first scraper matching the URL will be used
  - url: www.pornhub.com/model/
    action: scrapeXPath
    # must be a name defined under xPathScrapers
    scraper: performerScraper

# scraping rules for scraping scene by URL
sceneByURL:
  - url: www.pornhub.com/view_video
    action: scrapeXPath
    scraper: sceneScraper

# definitions of XPath scrapers
xPathScrapers:
  # these can be freely named and are referenced in the scraper field
  performerScraper:
    # define common prefixes that are reused
    common:
      # these must be prefixed with $ character
      $smallInfo: //div[@class="content-columns"]//span[@class="smallInfo"]
    performer:
      # these should be named the same as the performer fields
      # value is the xpath to match to get the value of the field
      # by default will get the text content
      name: //h1[@itemprop="name"]
      # these use the $smallInfo prefix defined in common
      birthdate: $smallInfo[2]
      measurements: $smallInfo[6]
      height: $smallInfo[7]
      ethnicity: $smallInfo[9]
      fake_tits: $smallInfo[12]
      tattoos: $smallInfo[13]
      career_length: $smallInfo[19]

  sceneScraper:
    common:
      $performerElement: //div[@class="video-info-row"]//div[@class="usernameWrap"]//a
    scene:
      title: //div[@id="main-container"][0]/@data-video-title
      performers:
        # these should be named the same as the performer fields
        name: $performerElement
        # this may allow scraping of the performer details
        url: $performerElement/@href
      tags: //div[@class="categoriesWrapper"]//a[not(text() = "+ Suggest")]
      date: 
        # this field requires post-processing, so the selector is defined separately
        selector: //script[@type="application/ld+json"]
        # post-process step
        replace: 
          # each regex is performed in order
          - regex: '.*"uploadDate": "(.*?)",.*'
            # $1 refers to the first capture group
            with: $1

```

My rough plan for this is to open PRs for porting the existing format for performers first, then to adjust the scene scraping PR to accomodate the YAML format, then a final PR to implement the XPath scraping functionality.

Let me know what you think.

**Comment by WithoutPants** [comment]
Once #333 is merged, I think this can be closed as complete.
